# Лабораторная работа 4 по криптографии
## Критерий открытого текста

*Выполнил Моисеенков Илья Павлович, М8О-308Б-19*

### Задание
Сравнить:  
1) два осмысленных текста на естественном языке  
2) осмысленный текст и текст из случайных букв  
3) осмысленный текст и текст из случайных слов  
4) два текста из случайных букв  
5) два текста из случайных слов.

Как сравнивать: считать процент совпадения букв в сравниваемых текстах – получить дробное значение от 0 до 1 как результат деления количества совпадений на общее число букв. Расписать подробно в отчёте алгоритм сравнения и приложить сравниваемые тексты в отчёте хотя бы для одного запуска по всем пяти подпунктам. Осознать какие значения получаются в этих пяти подпунктах. Привести свои соображения о том почему так происходит.

Длина сравниваемых текстов должна совпадать. Привести соображения о том какой длины текста должно быть достаточно для корректного сравнения.

---

Для работы я взял текст Войны и мира на английском языке. Я приведу весь текст к нижнему регистру и уберу пунктуацию.

In [1]:
import string

with open('war_and_peace.txt') as f:
    full_text_with_punctuation = f.read().replace('\n', ' ').replace('\r', '').lower()
    punctuation = set(string.punctuation)
    full_text = ''.join(c for c in full_text_with_punctuation if c not in punctuation)

In [2]:
len(full_text)

3089536

Возьмем в качестве первого текста первые *text_len* символов, а в качестве второго текста последние *text_len* символов.

In [3]:
text_len = 100000

In [4]:
text1 = full_text[:text_len]
text2 = full_text[-text_len:]

In [5]:
assert len(text1) == len(text2)

In [6]:
text1

'chapter i  well prince so genoa and lucca are now just family estates of the buonapartes but i warn you if you dont tell me that this means war if you still try to defend the infamies and horrors perpetrated by that antichristi really believe he is antichristi will have nothing more to do with you and you are no longer my friend no longer my faithful slave as you call yourself but how do you do i see i have frightened yousit down and tell me all the news  it was in july 1805 and the speaker was the wellknown anna pavlovna scherer maid of honor and favorite of the empress marya fedorovna with these words she greeted prince vasili kuragin a man of high rank and importance who was the first to arrive at her reception anna pavlovna had had a cough for some days she was as she said suffering from la grippe grippe being then a new word in st petersburg used only by the elite  all her invitations without exception written in french and delivered by a scarletliveried footman that morning ran 

In [7]:
text2

'tions act as they wished and by what was the will of these individuals themselves guided the ancients met by recognizing a divinity which subjected the nations to the will of a chosen man and guided the will of that chosen man so as to accomplish ends that were predestined  for the ancients these questions were solved by a belief in the direct participation of the deity in human affairs  modern history in theory rejects both these principles  it would seem that having rejected the belief of the ancients in mans subjection to the deity and in a predetermined aim toward which nations are led modern history should study not the manifestations of power but the causes that produce it but modern history has not done this having in theory rejected the view held by the ancients it still follows them in practice  instead of men endowed with divine authority and directly guided by the will of god modern history has given us either heroes endowed with extraordinary superhuman capacities or simpl

### Генерация рандомных текстов

Напишем функцию для генерации текстов из рандомных букв и слов.

Для генерации из букв будем использовать только английские буквы нижнего регистра и пробел.  
Для генерации из слов будем использовать все слова английского языка, разделяя их пробелами. Если последнее слово оказалось длиннее, чем нужно, то обрежем его.

In [8]:
import numpy as np
import random

In [9]:
def get_random_chars_text(size):
    """
    Generate text with random chars and spaces with given len
    """
    letters = string.ascii_lowercase + ' '

    random_chars_text = np.random.choice(list(letters), size, replace=True)
    random_chars_text = ''.join(random_chars_text)
    assert len(random_chars_text) == size
    return random_chars_text

In [10]:
get_random_chars_text(100)

'wyqm oitbsnnvqywppdmqlspicoevbzesgyqdkpbyhxfecsmbqhkahzzhlkblvxq xavxwxgen gmy ltzorvrqpveowzlopbqaq'

In [11]:
with open('english_words.txt') as f:
    english_words = [word.strip() for word in f.readlines()]

In [12]:
def get_random_words_text(size):
    """
    Generate text with random words with given len
    """
    
    cur_len = 0
    words_for_text = []
    while cur_len <= size:
        random_word = random.choice(english_words)
        cur_len += len(random_word) + 1
        words_for_text.append(random_word)
    
    result = ' '.join(words_for_text)[:size]
    assert len(result) == size
    return result

In [13]:
get_random_words_text(100)

'intermigrate indeclinably madid majorem muscatorium treponemicide ciruela yengees yrs yawnproof nipp'

### Сравнение текстов

Напишем функцию для сравнения текстов. Будем использовать метрику accuracy

In [14]:
def accuracy(text1, text2):
    assert len(text1) == len(text2)
    matched = 0
    for c1, c2 in zip(text1, text2):
        if c1 == c2:
            matched += 1
    
    return matched / len(text1) * 100

#### 1. Сравниваем два нормальных текста

In [15]:
print(f'Точность при сравнении осмысленных текстов = {accuracy(text1, text2)}%')

Точность при сравнении осмысленных текстов = 7.698%


#### 2. Сравниваем нормальный текст с текстом из случайных букв

In [16]:
print(f'Точность при сравнении осмысленного текста и текста из рандомных букв = {accuracy(get_random_chars_text(len(text1)), text2)}%')

Точность при сравнении осмысленного текста и текста из рандомных букв = 3.739%


#### 3. Сравниваем нормальный текст и текст из случайных слов

In [17]:
print(f'Точность при сравнении осмысленного текста и текста из рандомных слов = {accuracy(get_random_words_text(len(text2)), text2)}%')

Точность при сравнении осмысленного текста и текста из рандомных слов = 6.313000000000001%


#### 4. Сравниваем два текста из рандомных букв

In [18]:
print(f'Точность при сравнении двух текстов из рандомных букв = {accuracy(get_random_chars_text(len(text1)), get_random_chars_text(len(text2)))}%')

Точность при сравнении двух текстов из рандомных букв = 3.7379999999999995%


#### 5. Сравниваем два текста из рандомных слов

In [19]:
print(f'Точность при сравнении двух текстов из рандомных слов = {accuracy(get_random_words_text(len(text2)), get_random_words_text(len(text2)))}%')

Точность при сравнении двух текстов из рандомных слов = 5.976999999999999%


### Вывод

Выполнив данную лабораторную работу, я получил следующие результаты:  
* точность при сравнении двух нормальных текстов = ~7%  
* точность при сравнении текста из рандомных слов с чем-нибудь = ~6%  
* точность при сравнении текста из рандомных букв с чем-нибудь = ~3.7%

Видим, что при использовании слов (даже рандомных) точность совпадений выше, чем при использовании рандомных букв. В словах буквы связаны друг с другом: в них встречаются частые буквосочетания, частые паттерны итд. В тексте из рандомных букв между символами нет абсолютно никакой связи. Мы можем получить абсурдные сочетания букв и слов, которые маловероятно встретятся в реальном тексте.

Но точность при сравнении текстов из рандомных слов и реальных текстов тоже отличается. Тексты из слов не учитывают связей между этими словами. А реальные тексты, очевидно, учитывают.

При помощи таких оценок мы можем примерно понять, какой текст мы получили (допустим, он зашифрован и мы не можем просто посмотреть на него глазами, или он на незнакомом языке, или он слишком большой и мы не можем изучить его полностью). Мы можем посчитать точности совпадений с реальными текстами и посмотреть на полученное значение. По этому показателю мы сможем понимать, написано ли в тексте что-то осмысленное или это просто набор букв. А если точность околонулевая, то возможно текст вообще на другом языке. Такие проверки используются в некоторых алгоритмах криптографии.

Выполняя работу, я пробовал рассматривать тексты разной длины: 10000, 50000, 100000 символов. При всех значениях я получал примерно одинаковый результат. Поэтому этот метод будет работать для любых достаточно больших текстов.